zebra-chain/src/orchard/sinsemilla.rs

```diff
@@ -0,0 +1,134 @@
+//! Sinsemilla hash functions and helpers.
+
+use bitvec::prelude::*;
+
+use halo2::{
+    arithmetic::{CurveAffine, CurveExt},
+    pasta::pallas,
+};
+
+/// [Hash Extractor for Pallas][concreteextractorpallas]
+///
+/// P → B^[l^Orchard_Merkle]
+///
+/// [concreteextractorpallas]: https://zips.z.cash/protocol/nu5.pdf#concreteextractorpallas
+// TODO: should this return the basefield element type, or the bytes?
+pub fn extract_p(point: pallas::Point) -> pallas::Base {
+    match pallas::Affine::from(point).get_xy().into() {
+        // If Some, it's not the identity.
+        Some((x, _)) => x,
+        _ => pallas::Base::zero(),
+    }
+}
+
+/// GroupHash into Pallas, aka _GroupHash^P_
+///
+/// Produces a random point in the Pallas curve.  The first input element acts
+/// as a domain separator to distinguish uses of the group hash for different
+/// purposes; the second input element is the message.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretegrouphashpallasandvesta
+#[allow(non_snake_case)]
+pub fn pallas_group_hash(D: &[u8], M: &[u8]) -> pallas::Point {
+    let domain_separator = std::str::from_utf8(D).unwrap();
+
+    pallas::Point::hash_to_curve(domain_separator)(M)
+}
+
+/// Q(D) := GroupHash^P(︀“z.cash:SinsemillaQ”, D)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+fn Q(D: &[u8]) -> pallas::Point {
+    pallas_group_hash(b"z.cash:SinsemillaQ", D)
+}
+
+/// S(j) := GroupHash^P(︀“z.cash:SinsemillaS”, LEBS2OSP32(I2LEBSP32(j)))
+///
+/// S: {0 .. 2^k - 1} -> P^*, aka 10 bits hashed into the group
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+// XXX: should j be strictly limited to k=10 bits?
```

Comment:
Yeah we probably should limit `j` to 10 bits, can we use a `BitVec`?

Reply:
Resolved in c1dfeb7a